// rc-banner-anim/lib/Arrow.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

var Arrow = function (_Component) {
  _inherits(Arrow, _Component);

  function Arrow() {
    _classCallCheck(this, Arrow);

    var _this = _possibleConstructorReturn(this, _Component.apply(this, arguments));

    ['onClick'].forEach(function (method) {
      return _this[method] = _this[method].bind(_this);
    });
    return _this;
  }

  Arrow.prototype.onClick = function onClick() {
    this.props[this.props.arrowType]();
  };

  Arrow.prototype.render = function render() {
    var className = this.props.className;
    var defaultClass = className + '-default';
    className = (className + ' ' + (this.props.prefixCls || '')).trim();
    className = !this.props["default"] ? className : (className + ' ' + defaultClass).trim();
    className = className + ' ' + this.props.arrowType;
    var props = (0, _objectAssign2["default"])({}, this.props);
    ['arrowType', 'next', 'prev', 'elemHeight', 'component', 'default', 'prefixCls'].forEach(function (key) {
      return delete props[key];
    });
    props.className = className;
    props.onClick = this.onClick;
    props.style = props.style || {};
    props.style.top = this.props.elemHeight / 2 + 'px';
    return _react2["default"].createElement(this.props.component, props, this.props.children);
  };

  return Arrow;
}(_react.Component);

Arrow.propTypes = {
  children: _react.PropTypes.any,
  style: _react.PropTypes.object,
  className: _react.PropTypes.string,
  prefixCls: _react.PropTypes.string,
  component: _react.PropTypes.any,
  arrowType: _react.PropTypes.string,
  "default": _react.PropTypes.bool,
  next: _react.PropTypes.func,
  prev: _react.PropTypes.func,
  elemHeight: _react.PropTypes.number
};
Arrow.defaultProps = {
  component: 'div',
  className: 'banner-anim-arrow'
};

exports["default"] = Arrow;
module.exports = exports['default'];
