// rc-banner-anim/lib/utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.toArrayChildren = toArrayChildren;
exports.dataToArray = dataToArray;
exports.setAnimCompToTagComp = setAnimCompToTagComp;
exports.currentScrollTop = currentScrollTop;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _rcTweenOne = require('rc-tween-one');

var _rcTweenOne2 = _interopRequireDefault(_rcTweenOne);

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function toArrayChildren(children) {
  var ret = [];
  _react2["default"].Children.forEach(children, function (c) {
    ret.push(c);
  });
  return ret;
}

function dataToArray(vars) {
  if (!vars && vars !== 0) {
    return [];
  }
  if (Array.isArray(vars)) {
    return vars;
  }
  return [vars];
}

function setAnimCompToTagComp(item, i) {
  if (!item) {
    return null;
  }
  var props = (0, _objectAssign2["default"])({}, item.props);
  props.key = item.key || i;
  // 压缩后名称不一样了。
  var propTypes = item.type.propTypes;
  if (propTypes && (propTypes.animConfig && propTypes.animatingClassName && propTypes.leaveReverse && propTypes.delay && propTypes.ease && propTypes.interval && propTypes.duration || item.type === _rcTweenOne2["default"] || propTypes.showProp && propTypes.exclusive && propTypes.transitionName && propTypes.transitionAppear && propTypes.transitionEnter && propTypes.transitionLeave && propTypes.onEnd && propTypes.animation || props.name === 'QueueAnim' || props.name === 'TweenOne' || props.name === 'Animate')) {
    // queueAnim or tweeOne or animate;
    var style = (0, _objectAssign2["default"])({}, props.style);
    style.position = 'relative';
    props.style = style;
    var component = props.component;
    ['component', 'interval', 'duration', 'delay', 'animConfig', 'ease', 'leaveReverse', 'animatingClassName', 'animation', 'reverseDelay', 'attr', 'showProp', 'exclusive', 'transitionName', 'transitionAppear', 'transitionEnter', 'transitionLeave', 'onEnd'].forEach(function (key) {
      return delete props[key];
    });
    return _react2["default"].createElement(component, props);
  }
  return item;
}
setAnimCompToTagComp.propTypes = {
  key: _react2["default"].PropTypes.string,
  style: _react2["default"].PropTypes.object,
  component: _react2["default"].PropTypes.any,
  name: _react2["default"].PropTypes.string
};

function currentScrollTop() {
  var supportPageOffset = window.pageXOffset !== undefined;
  var isCSS1Compat = (document.compatMode || '') === 'CSS1Compat';
  var isCSS1ScrollTop = isCSS1Compat ? document.documentElement.scrollTop : document.body.scrollTop;
  return supportPageOffset ? window.pageYOffset : isCSS1ScrollTop;
}
