// rc-tween-one/lib/ticker.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _raf = require('raf');

var _raf2 = _interopRequireDefault(_raf);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var Ticker = function Ticker() {}; /* eslint-disable func-names */

var p = Ticker.prototype = {
  tickFnObject: {},
  id: -1,
  autoSleep: 120,
  frame: 0,
  perFrame: Math.round(1000 / 60)
};
p.wake = function (key, fn) {
  this.tickFnObject[key] = fn;
  if (this.id === -1) {
    this.id = (0, _raf2["default"])(this.tick);
  }
};
p.clear = function (key) {
  delete this.tickFnObject[key];
};
p.sleep = function () {
  _raf2["default"].cancel(this.id);
  this.id = -1;
};
var ticker = new Ticker();
p.tick = function (a) {
  var obj = ticker.tickFnObject;
  Object.keys(obj).forEach(function (key) {
    if (obj[key]) {
      obj[key](a);
    }
  });
  // 如果 object 里没对象了，自动睡眠；
  if (!Object.keys(obj).length) {
    return ticker.sleep();
  }
  ticker.frame++;
  ticker.id = (0, _raf2["default"])(ticker.tick);
};
var timeoutIdNumber = 0;
p.timeout = function (fn, time) {
  var _this = this;

  if (!(typeof fn === 'function')) {
    return console.warn('Is no function'); // eslint-disable-line
  }
  var timeoutID = 'timeout' + Date.now() + '-' + timeoutIdNumber;
  var startFrame = this.frame;
  this.wake(timeoutID, function () {
    var moment = (_this.frame - startFrame) * _this.perFrame;
    if (moment >= (time || 0)) {
      _this.clear(timeoutID);
      fn();
    }
  });
  timeoutIdNumber++;
  return timeoutID;
};
var intervalIdNumber = 0;
p.interval = function (fn, time) {
  var _this2 = this;

  if (!(typeof fn === 'function')) {
    console.warn('Is no function'); // eslint-disable-line
    return null;
  }
  var intervalID = 'interval' + Date.now() + '-' + intervalIdNumber;
  var starFrame = this.frame;
  this.wake(intervalID, function () {
    var moment = (_this2.frame - starFrame) * _this2.perFrame;
    if (moment >= (time || 0)) {
      starFrame = _this2.frame;
      fn();
    }
  });
  intervalIdNumber++;
  return intervalID;
};
exports["default"] = ticker;
module.exports = exports['default'];
